Add GLX_EXT_no_config_context

nwnk opened the pull request:
There is significant interest in enabling on-screen HDR drawables under
X11. One awkward feature of GLX_ARB_fbconfig_float is that, since the
context and fbconfig must be compatible, a compositor wanting to consume
both integer and floating point drawables must use multiple contexts.
To address this, port EGL_KHR_no_config_context to GLX.

oddhack commented:
This is in response to the X.org conference presentation?

Probably  good to add GLX protocol before it's registered.

I'm looking forward to an HDR HDPI UHD 43+" monitor, one of these days. Preferably with PoE, USB ports, and a few more acronyms attached to it as well.

nwnk commented:
Yes, this came up in that talk and the ensuing hallway discussion. It's probably not all that will be needed for HDR GLX windows, the fbconfig attributes may need to be carefully specified to avoid spoiling non-HDR-aware clients, but this will certainly make the compositor's life easier and seemed like useful functionality on its own.

Updated spec makes the protocol encoding explicit and covers a case I missed in glXQueryContext.

oddhack commented:
Thanks. I'll rely on you to get the appropriate degree of EXT signoff from Mesa / NVIDIA / AMD / etc. and remove the WIP when it's ready.

nwnk commented:
Went ahead and implemented this for [Mesa](https://lists.freedesktop.org/archives/mesa-dev/2017-September/170374.html) and [Xorg](https://lists.freedesktop.org/archives/xorg-devel/2017-September/054677.html), discovered in the process that only CreateContextAttribs is expressive enough to support this. Updated the draft to match.

cubanismo commented:
I was going to suggest this during the talk, but then forgot the name of the EGL extension.  Looks like you're way ahead of me.

@AlexGoinsNV will probably want to review this as well.

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+    RESOLVED: This is an implementation detail rather than a spec issue.
+    glReadBuffer/glDrawBuffer have undefined results if called without a
+    current context. The GL_DRAW_BUFFER and GL_READ_BUFFER are set on the
```
s/current context/current drawable/?

nwnk replied:
(another copypasta from EGL but) no, I think this is right. Without a current context, GL calls are undefined. glDrawBuffer with a current context (that may have had null drawables named when bound) merely names an FBO to draw to, which need not be one provided by the window system.

The question here is "what is the initial state of GL_DRAW_BUFFER and GL_READ_BUFFER for the default framebuffer?", and I think we could reasonably say "undefined, name it if you want it". Certainly that's what would have to happen in Mesa for your drawing to go anywhere. I agree it's a spec issue, in that the WS binding should be clear about it, but nobody can possibly be expecting anything meaningful to happen if you don't name a real draw/read buffer, and the WS's spec should say as much.

cubanismo replied:
Agreed, I think I hit submit on this one before reading it through all the way.

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    first glXMakeCurrent call and can be updated in glReadBuffer and
+    glDrawBuffers calls after that. Therefore, the attribute value with
+    which the context is created is irrelevant from the point of view of the
+    spec and is left up to the implementation.
```
I had a hell of a time getting this right when our implementation add no_surface support.  It's definitely a spec issue, I just don't think it's an issue for this spec to tackle.  Otherwise, what you've said all makes sense.  The state of the GL portion of a GLX context before its been made current is generally akin to the "If a tree falls in the woods..." thought exercise.  It only becomes interesting when some jerk calls glXCopyContext() on your context before it has been made current IIRC.

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    NULL, GL_BACK if the surface is double buffered, GL_FRONT if the surface
+    is single buffered. Following calls to glReadBuffer and glDrawBuffers
+    change the GL_DRAW_BUFFER and GL_READ_BUFFER attributes and these values
+    persist even when the application change the current context.
```
While we're pointing out weird spec quirks, you might want to refer to the extensive language GLX has covering what happens if the 2nd drawable a context is made current to doesn't contain the current read or draw buffer referred to by the context.  See page 28 of GLX 1.4.

nwnk replied:
This is already addressed in the GLX_ARB_create_context spec, I don't think this extension has anything to add there.

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    change the GL_DRAW_BUFFER and GL_READ_BUFFER attributes and these values
+    persist even when the application change the current context.
+
+ 5) Should we add an glXCreateGenericContext which is the same as
```
s/an/a/

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+    RESOLVED: No.
+
+ 6) Can no-config contexts share state with contexts that has a config?
```
Should we require the version of each context is sufficient to support surfaceless make current at least?  Also, s/that has/that have/

cubanismo commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+ 7) What surface types can be made current with a no-config context?
+
+    RESOLVED: any surface type supported by the implementation can be made
```
I'm assuming by supported, you mean which can be created according to the FBConfigs it exports, not that the implementation can arbitrarily reject random drawables it doesn't like here.  Also, s/surface type/GLXDrawable types/.  Surface is an EGLism.

oddhack commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+    "If the context was created without reference to a config, the value
+    of GLX_RENDER_TYPE will be GLX_DONT_CARE, and the values of
+    GLX_FBCONFIG_ID and GLX_VISUAL_ID_EXT will be GLX_NO_CONFIG_EXT."
```
GLXFBConfig is a pointer, not an XID. Not sure if there's precedent for assigning None to a pointer? It's actually defined to be '0L' in the X.h on my machine.

kbrenneman replied:
Using NULL should be fine. That's what it uses for GLXContext already, which is the other opaque pointer type in GLX.

cubanismo replied:
GLXFBConfig is a pointer, but GLX_FBCONFIG_ID is the XID of an FBConfig.  E.g., an app can query the value of GLX_FBCONFIG_ID from a GLXFBConfig using glXGetFBConfigAttrib().

kbrenneman commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    "If <config> is GLX_NO_CONFIG_EXT, the context will be created without
+    reference to a config. In this case, the attribute list must include
+    GLX_SCREEN, naming the screen for the contexts; if GLX_SCREEN is not
+    specified, or GLX_RENDER_TYPE is specified, BadValue is generated."
```
If we implement PRIME offloading in GLVND in terms of a ((screen, client) -> vendor) mapping, then we'd want a screen number, not a drawable. Depending on how the server handles XID's in that case (using the client that sent a request versus the client that owns the XID), using a drawable could lead to unexpected vendor mismatches.

Besides that, if the resulting context is supposed to be compatible with any drawable on the same screen, then using a screen number is unambiguous.

kbrenneman commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    which can support them?
+
+    RESOLVED: Yes. ES2 + OES_surfaceless_context, ES 3.0, and GL 3.0 all
+    support binding a surface without a context. This implies that they
```
Shouldn't that be "binding a context without a surface"? Or I guess for GLX, "binding a context without a drawable"?

kbrenneman commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+    and the <screen> field must match, or BadValue is generated. This implies
+    that the client library is responsible for pulling the screen number out
+    of the attribute list, since glXCreateContextAttribsARB does not have an
+    explicit screen parameter.
```
The client side should be simple enough either way. We'll probably need to add a dispatch stub for glXCreateContextAttribs to GLVND itself, though. Right now, it's a vendor-provided stub, and while a vendor could provide a stub that works with this extension, you could end up with different results depending on which vendor libGLX.so actually gets that stub from.

As for the protocol, if it turns out we do need to include a drawable, then we could just include it in the attribute list in the request.

nwnk commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+    Accepted as the <config> parameter of glXCreateContextAttribsARB:
+
+        GLX_NO_CONFIG_EXT                   ((GLXFBConfig)0)
```
Totally fine with just using NULL, means I don't have to embed that cast in the otherwise language-neutral XML.

nwnk commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+ 1) Should non-conformant no-config contexts be allowed to be created?
+
+    RESOLVED: No. We are not encouraging non-conformant contexts.
```
This was just copypasta from the EGL spec, I have no problem deleting it. Other GLXes have had non-conformant caveats on the ARGB visual, which is in some sense mandatory for modern X, so this would more be saying "we don't have a good definition for conformance here" than encouraging bad GL, I think.

nwnk commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,254 @@
+
+    RESOLVED: Yes. ES2 + OES_surfaceless_context, ES 3.0, and GL 3.0 all
+    support binding a surface without a context. This implies that they
+    don't need to know surface attributes at context creation time.
```
Again, just copied from the EGL spec. I do have some interest in indirect GLES2, as it seems to be both a) what toolkits for compositors have as an alternate direct-context target, b) simpler than full indirect GL 2.0 in terms of implementation effort. But this particular language is just saying that if you ask for a pre-classic-3.0 context without reference to an fbconfig, you have no reason to expect success.

In an implementation sense I don't think Mesa has ever supported any targets where we would really _need_ to enforce this at context creation, so if someone wanted to support this on a GL{,ES}1 context then this restriction does seem unnecessary. I don't know of anybody retrocomputing that hard, but sure let's not make their lives any harder than it already is.

nwnk commented:
Updated from review comments, hopefully I addressed everything.

The trickiest changes here are around the non-conformance and share-context issues. For the former I've merely required that the context be conformant when bound to a conformant-config drawable, or to no drawable, which seems to be the original intent. For the latter the behavior is as if the two contexts had the same config; GLX_ARB_create_context already allows significant implementation freedom if context attributes don't match.

I tend to agree with @kbrenneman about the screen number issues. Whether it's good X extension style or not, it's what GLX already does.

nwnk commented on extensions/EXT/GLX_EXT_no_config_context.txt:
```diff
@@ -0,0 +1,289 @@
+
+    A context created without reference to a config may be used to render to
+    drawables of GLX_RGBA_UNSIGNED_FLOAT_TYPE_EXT.
+
```
Maybe it'd be useful? It's less precision than 10bpc but wider range, and in either case it's half the memory of FP16 so there might be a performance reason to want it. On the other hand I know 10bpc is a display format, but I don't know of any hardware that can display packed-float, so you'd always need at least one conversion before presentation.

The rgb9e5 shared-exponent format seems like it'd be more directly suitable, but there I'm not even aware of any hardware where that's even a renderable format let alone displayable.

AlexGoinsNV replied:
I guess it wouldn't hurt to include it, given that EXT_no_config_context allows for it.

Agreed that the shared-exponent format seems more suitable for preserving HDR content, but as long as it's really only useful for e.g. texture/image storage probably doesn't make sense to work towards render/display at an API level.

nwnk commented:
Updated the float fbconfig language from "drawables of FOO_TYPE" to "drawables whose config has FOO_BIT set".

nwnk commented:
@cubanismo did you have changes you still wanted to see here?

nwnk commented:
The Mesa half of this is merged now. The Xorg half has been reviewed, pending approval of this spec. Are there any further changes we need here?

nwnk commented:
The Mesa and Xorg implementations here uncovered a few corner cases. I'm not sure if they warrant mention in the spec itself, but I'll note them here for posterity at least.

On the client side, if you make a no-config context current against a drawable created by some other client, you'll need to do GLXGetDrawableAttributes to look up the appropriate fbconfig. If your GLX implementation descends from the SGI SI, you may need to ensure that GLX_FBCONFIG_ID is actually sent by the server and that the client library is prepared to accept it (the open source SI code drop did not implement GLX_SGIX_fbconfig / GLX 1.3).

On the server side, if a no-config context is made current against a Window (not a GLXWindow) then the server will need to decide which fbconfig to apply to the window. This was true of config-ful contexts already, but there you could simply apply the context's config. In principle multiple fbconfigs can point at the same visual. If so, the server would need to pick one; presumably one would use the same heuristic as one uses for glXCreateContext (i.e. creating the context with respect to a Visual not an fbconfig). That's all a bit theoretical, as Xorg's window-capable fbconfigs are bijective with its Visuals (and I believe the same is true of NVIDIA's GLX) so there's no ambiguity. This is app-visible in a sense, but I hesitate to require a particular behavior in the spec, particularly since the app is being a bit cheeky by using post-GLX-1.4 context creation but pre-GLX-1.3 drawables.

piglit has a touch test for this extension now, verifying that the MakeCurrent call actually works against every visual.

kbrenneman commented:
What we've got here should work with GLVND. I've written a dispatch stub for it (dispatching by the GLX_SCREEN attribute if there is one and the GLXFBConfig if there isn't). I still need to add unit tests for it before it's ready to check in.

Even with GLVND updated, though, we'll still want to update the vendor-provided stubs, too. That way, the extension still works (or at least, can work) if someone's using an older version of libglvnd.

Also, one other point in favor of specifying a screen in the attribute list instead of a drawable: If an app doesn't want to use a GLX drawable at all (e.g., it's going to do all of its drawing to an FBO), then it would make sense to also not require a GLXFBConfig when it creates the context. In that case, we shouldn't require the app to create a drawable that only exists to create a context.

cubanismo commented:
Sorry for the delay.  Looks like all the issues I raised were addressed, so LGTM.

nwnk commented:
Marked spec as complete, ready to merge.
